### PR TITLE
Feature : rewriting satisfaction form in twig

### PR DESCRIPTION
**Pull request by Lainow**
### <samp>🤖 Generated by Copilot at 889e1ae</samp>

Refactored the satisfaction form rendering for ITIL tickets to use the Twig template engine. Added a new Twig template file `ticketsatisfaction.html.twig` with the HTML code for the form and the rating widget.

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #number

### CODE REVIEW

File: templates/pages/management/ticketsatisfaction.html.twig [hunk 1]
```diff
@@ -0,0 +1,64 @@
+{% import "components/form/fields_macros.html.twig" as fields %}
```
Comment (stonebuzz) [inline templates/pages/management/ticketsatisfaction.html.twig:1]: ```ticketsatisfaction.html.twig``` need to be located in ```templates/components/itilobject```

File: templates/pages/management/ticketsatisfaction.html.twig [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+{% block more_fields %}
+    <div class='alert alert-success d-flex align-items-center' role='alert'>
+        <span class='ms-2'>
+        <i class='fas fa-xl fa-info-circle'> </i> &nbsp;
+            {{ __("After 12 hours, you can no longer modify your notice.") }}
+        </span>
+    </div>
```
Comment (stonebuzz) [inline templates/pages/management/ticketsatisfaction.html.twig:9]: Need to be added with ```fields.htmlField``` (to control width)

image

Green border should only on left side (see Lock tab from Computer)

image

**Review by stonebuzz — CHANGES_REQUESTED** [review]
From  ```showSatisactionForm``` (```CommonITILSatisfaction```) 
```php
if ((int) $this->fields["type"] === self::TYPE_EXTERNAL) {
            $url = Entity::generateLinkSatisfaction($item);
            echo "<div class='center spaced'>" .
                "<a href='$url'>" . __('External survey') . "</a><br>($url)</div>";
        } 
```
should no longer exist

**Comment by stonebuzz** [comment]
```{% extends "generic_show_form.html.twig" %}``` need to be used in twig template

to drop 

```php
            $options['candel'] = false;
            $this->showFormButtons($options);
``` 

from showSatisactionForm (CommonITILSatisfaction)

File: src/CommonITILSatisfaction.php [hunk 3]
```diff
@@ -133,66 +168,34 @@ public function canUpdateItem()
...
             $url = Entity::generateLinkSatisfaction($item);
-            echo "<div class='center spaced'>" .
-                "<a href='$url'>" . __('External survey') . "</a><br>($url)</div>";
+            TemplateRenderer::getInstance()->display('/components/itilobject/ticketsatisfaction.html.twig', [
+                'url' => $url
+            ]);
         } else { // for internal inquest => form
             $config_suffix = $item->getType() === 'Ticket' ? '' : ('_' . strtolower($item->getType()));
 
             $this->showFormHeader($options);
```
Comment (stonebuzz) [inline src/CommonITILSatisfaction.php:184]: can be removed

**Review by stonebuzz — CHANGES_REQUESTED** [review]

File: src/CommonITILSatisfaction.php [hunk 4]
```diff
@@ -1,5 +1,40 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * ---------------------------------------------------------------------
+ */
```
Comment (cedric-anne) [inline src/CommonITILSatisfaction.php:34]: Duplicated header. The `use` instruction should be right before the `abstract class` declaration.

File: templates/components/itilobject/ticketsatisfaction.html.twig [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+                    </span>
+                </div>
+            </div>"
+            ,
+            '',
+            {
+                'input_class' : 'col-12',
+                'label_class' : ''
+            }
+        ) }}
```
Comment (cedric-anne) [inline templates/components/itilobject/ticketsatisfaction.html.twig:68]: Is it possible to write the HTML code without using the `fields.htmlField()` macro?

File: templates/components/itilobject/ticketsatisfaction.html.twig [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        {{ fields.htmlField(
+            '',
+            "<a href='"~ url ~"'>" ~ url ~ "</a><br></div>"
+            ,
+            __('External survey'),
+            {
+                'input_class' : 'col-xxl-9',
+                'label_class' : 'col-xxl-3'
+            }
+        ) }}
```
Comment (cedric-anne) [inline templates/components/itilobject/ticketsatisfaction.html.twig:47]: Is it possible to write the HTML code without using the `fields.htmlField()` macro?

Also, you have a `</div>` but no div opening.

File: templates/components/itilobject/ticketsatisfaction.html.twig [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+                    <option value='{{ i }}' {% if i == item.fields['satisfaction'] %} selected {% endif %}>{{ i }}</option>
+                {% endfor %}
+            </select>
+            <div class='rateit' id='stars'></div>
+        {% endset %}
+
+        {{ fields.htmlField(
+            'satisfaction',
+            select_dom,
+            __('Satisfaction with the resolution of the ') ~ itemtypename
```
Comment (cedric-anne) [inline templates/components/itilobject/ticketsatisfaction.html.twig:88]:
```suggestion
            __('Satisfaction with the resolution of the %s')|format(ticket.getTypeName(1))
```

File: templates/components/itilobject/ticketsatisfaction.html.twig
Comment (cedric-anne) [inline templates/components/itilobject/ticketsatisfaction.html.twig:1]: This file will be used also for `ChangeSatisfaction`. It should be renamed to `itilsatisfaction.html.twig`.

File: src/CommonITILSatisfaction.php [hunk 8]
```diff
@@ -133,66 +168,34 @@ public function canUpdateItem()
...
-                echo "<tr class='tab_bg_2'>";
-                echo "<td>" . __('Response date to the satisfaction survey') . "</td><td>";
-                echo Html::convDateTime($this->fields["date_answered"]) . "</td></tr>\n";
-            }
-
-            $options['candel'] = false;
-            $this->showFormButtons($options);
+            TemplateRenderer::getInstance()->display('/components/itilobject/ticketsatisfaction.html.twig', [
+                'item'   => $this,
+                'ticket' => $item,
```
Comment (cedric-anne) [inline src/CommonITILSatisfaction.php:193]: Can also be a change.
```suggestion
                'parent_item' => $item,
```

File: src/CommonITILSatisfaction.php [hunk 9]
```diff
@@ -133,66 +168,34 @@ public function canUpdateItem()
...
+                'foreigntype' => $item::getForeignKeyField(),
+                'itemtypename' => $item::getTypeName(1),
```
Comment (cedric-anne) [inline src/CommonITILSatisfaction.php:195]: Could be call directly from template, e.g. `{{ parent_item.getForeignKeyField() }}`.
```suggestion
```

File: templates/components/itilobject/ticketsatisfaction.html.twig [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        ) }}
+
+        {% set select_dom %}
+            <select id='satisfaction_data' name='satisfaction'>
+                {% for i in range(0, max_rate) %}
+                    <option value='{{ i }}' {% if i == item.fields['satisfaction'] %} selected {% endif %}>{{ i }}</option>
+                {% endfor %}
+            </select>
+            <div class='rateit' id='stars'></div>
+        {% endset %}
```
Comment (cedric-anne) [inline templates/components/itilobject/ticketsatisfaction.html.twig:83]: It could probably be rendered using the `dropdownArrayField` macro.

**Review by cedric-anne — CHANGES_REQUESTED** [review]

File: templates/components/itilobject/ticketsatisfaction.html.twig (hunk 5 above)
Comment (Lainow) [reply templates/components/itilobject/ticketsatisfaction.html.twig:68]: With htmlField, the alert occupies only the width of the form, whereas without htmlField, it occupies the entire width of the page.
image

File: templates/components/itilobject/ticketsatisfaction.html.twig (hunk 10 above)
Comment (Lainow) [reply templates/components/itilobject/ticketsatisfaction.html.twig:83]: It's possible but the stars disappear

**Review by stonebuzz — APPROVED** [review]

**Review by cedric-anne — CHANGES_REQUESTED** [review]

File: templates/components/itilobject/ticketsatisfaction.html.twig (hunk 5 above)
Comment (orthagh) [reply templates/components/itilobject/ticketsatisfaction.html.twig]: assign the html in a variable with `set` directive:

```twig
{% set foo %}
    <div">
        ...
    </div>
{% endset %}
```

and then pass the variable to `htmlField`

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**